File: apps/api/src/routes/criminal.ts
import type { FastifyInstance } from 'fastify'
import type { AppContext } from '../context.js'
import { requireCapability } from '../middleware/authorization.js'
import {
  createGangSchema,
  addGangMemberSchema,
  createOperationSchema,
  operationOutcomeSchema,
  registerContrabandSchema,
  seizeContrabandSchema,
  recordTradeSchema,
  stageRaidSchema,
  completeRaidSchema,
  abortRaidSchema,
} from '@atc/operations'
import {
  CriminalError,
  GangNotFoundError,
  GangValidationError,
  GangAlreadyExistsError,
  GangMemberNotFoundError,
  GangMemberAlreadyActiveError,
  GangOperationNotFoundError,
  GangOperationImmutableError,
  ContrabandNotFoundError,
  ContrabandAlreadySeizedError,
  RaidNotFoundError,
  RaidImmutableError,
  RaidAlreadyActiveError,
  BlackMarketTransactionNotFoundError,
} from '@atc/criminal-runtime'

function criminalErrorToResponse(err: CriminalError): { status: number; error: string; message: string } {
  if (err instanceof GangValidationError)            return { status: 400, error: 'GangValidation',            message: err.message }
  if (err instanceof GangOperationImmutableError)    return { status: 422, error: 'GangOperationImmutable',    message: err.message }
  if (err instanceof RaidImmutableError)             return { status: 422, error: 'RaidImmutable',             message: err.message }
  if (err instanceof GangAlreadyExistsError)         return { status: 409, error: 'GangAlreadyExists',         message: err.message }
  if (err instanceof GangMemberAlreadyActiveError)   return { status: 409, error: 'GangMemberAlreadyActive',   message: err.message }
  if (err instanceof ContrabandAlreadySeizedError)   return { status: 409, error: 'ContrabandAlreadySeized',   message: err.message }
  if (err instanceof RaidAlreadyActiveError)         return { status: 409, error: 'RaidAlreadyActive',         message: err.message }
  if (err instanceof GangNotFoundError)              return { status: 404, error: 'GangNotFound',              message: err.message }
  if (err instanceof GangMemberNotFoundError)        return { status: 404, error: 'GangMemberNotFound',        message: err.message }
  if (err instanceof GangOperationNotFoundError)     return { status: 404, error: 'GangOperationNotFound',     message: err.message }
  if (err instanceof ContrabandNotFoundError)        return { status: 404, error: 'ContrabandNotFound',        message: err.message }
  if (err instanceof RaidNotFoundError)              return { status: 404, error: 'RaidNotFound',              message: err.message }
  if (err instanceof BlackMarketTransactionNotFoundError) return { status: 404, error: 'BlackMarketTransactionNotFound', message: err.message }
  return { status: 500, error: 'CriminalError', message: err.message }
}

const NOT_CONFIGURED = { error: 'Criminal runtime not configured' }

export async function criminalRoutes(
  fastify: FastifyInstance,
  opts: { ctx: AppContext },
): Promise<void> {
  const { ctx } = opts

  // ── Create gang ───────────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/gangs', {
    preHandler: requireCapability(ctx, 'criminal:gang:manage'),
    handler: async (req, reply) => {
      if (!ctx.criminalRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = createGangSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const gang = await ctx.criminalRuntimeService.createGang(parsed.data)
        return reply.status(201).send(gang)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Get gang ──────────────────────────────────────────────────────────────────

  fastify.get('/api/v1/criminal/gangs/:gangId', {
    preHandler: requireCapability(ctx, 'criminal:gang:read'),
    handler: async (req, reply) => {
      if (!ctx.criminalRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const { gangId } = req.params as { gangId: string }
      const gang = await ctx.criminalRuntimeService.getGang(gangId)
      if (!gang) return reply.status(404).send({ error: 'GangNotFound', message: `Gang ${gangId} not found` })
      return reply.send(gang)
    },
  })

  // ── List active gangs ─────────────────────────────────────────────────────────

  fastify.get('/api/v1/criminal/gangs', {
    preHandler: requireCapability(ctx, 'criminal:gang:read'),
    handler: async (req, reply) => {
      if (!ctx.criminalRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const gangs = await ctx.criminalRuntimeService.listActiveGangs()
      return reply.send(gangs)
    },
  })

  // ── Add gang member ───────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/gangs/:gangId/members', {
    preHandler: requireCapability(ctx, 'criminal:gang:manage'),
    handler: async (req, reply) => {
      if (!ctx.criminalRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const { gangId } = req.params as { gangId: string }
      const parsed = addGangMemberSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const member = await ctx.criminalRuntimeService.addMember(gangId, parsed.data.principalId, parsed.data.rank, parsed.data.invitedByPrincipalId)
        return reply.status(201).send(member)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Remove gang member ────────────────────────────────────────────────────────

  fastify.delete('/api/v1/criminal/gangs/:gangId/members/:principalId', {
    preHandler: requireCapability(ctx, 'criminal:gang:manage'),
    handler: async (req, reply) => {
      if (!ctx.criminalRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const { gangId, principalId } = req.params as { gangId: string; principalId: string }
      try {
        const member = await ctx.criminalRuntimeService.removeMember(gangId, principalId)
        return reply.status(200).send(member)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Create operation ──────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/operations', {
    preHandler: requireCapability(ctx, 'criminal:operation:manage'),
    handler: async (req, reply) => {
      if (!ctx.gangOperationService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = createOperationSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const op = await ctx.gangOperationService.createOperation(parsed.data)
        return reply.status(201).send(op)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Start operation ───────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/operations/:operationId/start', {
    preHandler: requireCapability(ctx, 'criminal:operation:manage'),
    handler: async (req, reply) => {
      if (!ctx.gangOperationService) return reply.status(503).send(NOT_CONFIGURED)
      const { operationId } = req.params as { operationId: string }
      try {
        const op = await ctx.gangOperationService.startOperation(operationId)
        return reply.status(200).send(op)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Complete operation ────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/operations/:operationId/complete', {
    preHandler: requireCapability(ctx, 'criminal:operation:manage'),
    handler: async (req, reply) => {
      if (!ctx.gangOperationService) return reply.status(503).send(NOT_CONFIGURED)
      const { operationId } = req.params as { operationId: string }
      const parsed = operationOutcomeSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const op = await ctx.gangOperationService.completeOperation(operationId, parsed.data.outcome)
        return reply.status(200).send(op)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Abort operation ───────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/operations/:operationId/abort', {
    preHandler: requireCapability(ctx, 'criminal:operation:manage'),
    handler: async (req, reply) => {
      if (!ctx.gangOperationService) return reply.status(503).send(NOT_CONFIGURED)
      const { operationId } = req.params as { operationId: string }
      try {
        const op = await ctx.gangOperationService.abortOperation(operationId)
        return reply.status(200).send(op)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Register contraband ───────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/contraband', {
    preHandler: requireCapability(ctx, 'criminal:contraband:manage'),
    handler: async (req, reply) => {
      if (!ctx.contrabandService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = registerContrabandSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const contraband = await ctx.contrabandService.register(parsed.data)
        return reply.status(201).send(contraband)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Seize contraband ──────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/contraband/:contrabandId/seize', {
    preHandler: requireCapability(ctx, 'criminal:contraband:seize'),
    handler: async (req, reply) => {
      if (!ctx.contrabandService) return reply.status(503).send(NOT_CONFIGURED)
      const { contrabandId } = req.params as { contrabandId: string }
      const parsed = seizeContrabandSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const contraband = await ctx.contrabandService.seize(contrabandId, parsed.data.seizedByPrincipalId)
        return reply.status(200).send(contraband)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Record black market trade ─────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/trade', {
    preHandler: requireCapability(ctx, 'criminal:trade:record'),
    handler: async (req, reply) => {
      if (!ctx.blackMarketService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = recordTradeSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const trade = await ctx.blackMarketService.recordTrade(parsed.data)
        return reply.status(201).send(trade)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Stage raid ────────────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/raids', {
    preHandler: requireCapability(ctx, 'criminal:raid:manage'),
    handler: async (req, reply) => {
      if (!ctx.raidRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = stageRaidSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const raid = await ctx.raidRuntimeService.stageRaid(parsed.data)
        return reply.status(201).send(raid)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Start raid ────────────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/raids/:raidId/start', {
    preHandler: requireCapability(ctx, 'criminal:raid:manage'),
    handler: async (req, reply) => {
      if (!ctx.raidRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const { raidId } = req.params as { raidId: string }
      try {
        const raid = await ctx.raidRuntimeService.startRaid(raidId)
        return reply.status(200).send(raid)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Complete raid ─────────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/raids/:raidId/complete', {
    preHandler: requireCapability(ctx, 'criminal:raid:manage'),
    handler: async (req, reply) => {
      if (!ctx.raidRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const { raidId } = req.params as { raidId: string }
      const parsed = completeRaidSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const raid = await ctx.raidRuntimeService.completeRaid(raidId, parsed.data.outcome, parsed.data.notes)
        return reply.status(200).send(raid)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })

  // ── Abort raid ────────────────────────────────────────────────────────────────

  fastify.post('/api/v1/criminal/raids/:raidId/abort', {
    preHandler: requireCapability(ctx, 'criminal:raid:manage'),
    handler: async (req, reply) => {
      if (!ctx.raidRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const { raidId } = req.params as { raidId: string }
      const parsed = abortRaidSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const raid = await ctx.raidRuntimeService.abortRaid(raidId, parsed.data.notes)
        return reply.status(200).send(raid)
      } catch (err) {
        if (err instanceof CriminalError) return reply.status(criminalErrorToResponse(err).status).send(criminalErrorToResponse(err))
        throw err
      }
    },
  })
}

File: game/atc-core/server/criminal.lua
-- ATC Criminal Runtime Bridge — gang operations, raids, contraband, black market
-- All principal IDs resolved server-side. No client trust for operation state.

local API_BASE  = ATC.Config.ApiBase  or 'http://localhost:3000'
local API_TOKEN = ATC.Config.ApiToken or ''

local function apiPost(path, body, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'POST', json.encode(body), {
    ['Content-Type']  = 'application/json',
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

local function apiGet(path, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'GET', '', {
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

local function apiDelete(path, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'DELETE', '', {
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

-- ── Public SDK ────────────────────────────────────────────────────────────────

ATC.Criminal = {}

--- Create a new gang.
--- @param source          number   leader's FiveM server id
--- @param name            string
--- @param tag             string   max 8 chars
--- @param cb              function callback(status, gang)
function ATC.Criminal.CreateGang(source, name, tag, cb)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then
    if cb then cb(403, nil) end
    return
  end
  apiPost('/api/v1/criminal/gangs', {
    name               = name,
    tag                = tag,
    leaderPrincipalId  = principalId,
  }, cb)
end

--- Add a member to a gang.
--- @param source          number   inviting player's server id (capability check)
--- @param gangId          string
--- @param targetSource    number   player to add
--- @param rank            string   'leader'|'officer'|'member'|'associate'
--- @param cb              function callback(status, member)
function ATC.Criminal.AddMember(source, gangId, targetSource, rank, cb)
  local inviterPrincipalId = ATC.Accounts.GetPrincipalId(source)
  if not inviterPrincipalId then
    if cb then cb(403, nil) end
    return
  end
  local targetPrincipalId = ATC.Accounts.GetPrincipalId(targetSource)
  if not targetPrincipalId then
    if cb then cb(400, nil) end
    return
  end
  apiPost('/api/v1/criminal/gangs/' .. gangId .. '/members', {
    principalId          = targetPrincipalId,
    rank                 = rank or 'associate',
    invitedByPrincipalId = inviterPrincipalId,
  }, cb)
end

--- Remove a member from a gang.
--- @param source          number   officer/leader's server id
--- @param gangId          string
--- @param targetSource    number
--- @param cb              function callback(status, member)
function ATC.Criminal.RemoveMember(source, gangId, targetSource, cb)
  local targetPrincipalId = ATC.Accounts.GetPrincipalId(targetSource)
  if not targetPrincipalId then
    if cb then cb(400, nil) end
    return
  end
  apiDelete('/api/v1/criminal/gangs/' .. gangId .. '/members/' .. targetPrincipalId, cb)
end

--- Start a criminal operation.
--- @param source          number   operation owner
--- @param params          table    { label, operationType, gangId? }
--- @param cb              function callback(status, operation)
function ATC.Criminal.StartOperation(source, params, cb)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then
    if cb then cb(403, nil) end
    return
  end
  params = params or {}
  apiPost('/api/v1/criminal/operations', {
    label             = params.label          or 'Operation',
    operationType     = params.operationType  or 'other',
    ownerPrincipalId  = principalId,
    gangId            = params.gangId,
  }, cb)
end

--- Complete a criminal operation.
--- @param operationId     string
--- @param outcome         string   optional outcome description
--- @param cb              function callback(status, operation)
function ATC.Criminal.CompleteOperation(operationId, outcome, cb)
  apiPost('/api/v1/criminal/operations/' .. operationId .. '/complete', {
    outcome = outcome,
  }, cb)
end

--- Abort a criminal operation.
--- @param operationId     string
--- @param cb              function callback(status, operation)
function ATC.Criminal.AbortOperation(operationId, cb)
  apiPost('/api/v1/criminal/operations/' .. operationId .. '/abort', {}, cb)
end

--- Register contraband in a property stash.
--- @param source          number   registering player
--- @param params          table    { itemName, quantity, propertyId?, stashId? }
--- @param cb              function callback(status, contraband)
function ATC.Criminal.RegisterContraband(source, params, cb)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then
    if cb then cb(403, nil) end
    return
  end
  params = params or {}
  apiPost('/api/v1/criminal/contraband', {
    itemName                = params.itemName  or '',
    quantity                = tonumber(params.quantity) or 1,
    registeredByPrincipalId = principalId,
    propertyId              = params.propertyId,
    stashId                 = params.stashId,
  }, cb)
end

--- Seize contraband (law enforcement).
--- @param source          number   seizing officer
--- @param contrabandId    string
--- @param cb              function callback(status, contraband)
function ATC.Criminal.SeizeContraband(source, contrabandId, cb)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then
    if cb then cb(403, nil) end
    return
  end
  apiPost('/api/v1/criminal/contraband/' .. contrabandId .. '/seize', {
    seizedByPrincipalId = principalId,
  }, cb)
end

--- Record an illegal trade transaction.
--- @param sellerSource    number
--- @param buyerSource     number
--- @param params          table    { itemName, quantity, price, locationLabel? }
--- @param cb              function callback(status, transaction)
function ATC.Criminal.RecordTrade(sellerSource, buyerSource, params, cb)
  local sellerPrincipalId = ATC.Accounts.GetPrincipalId(sellerSource)
  if not sellerPrincipalId then
    if cb then cb(403, nil) end
    return
  end
  local buyerPrincipalId = ATC.Accounts.GetPrincipalId(buyerSource)
  if not buyerPrincipalId then
    if cb then cb(400, nil) end
    return
  end
  params = params or {}
  apiPost('/api/v1/criminal/trade', {
    sellerPrincipalId = sellerPrincipalId,
    buyerPrincipalId  = buyerPrincipalId,
    itemName          = params.itemName or '',
    quantity          = tonumber(params.quantity) or 1,
    price             = tonumber(params.price)    or 0,
    locationLabel     = params.locationLabel,
  }, cb)
end

--- Stage a raid on a property.
--- @param source          number   lead officer
--- @param params          table    { propertyId, initiatingAgencyId?, participants (table of sources), notes? }
--- @param cb              function callback(status, raid)
function ATC.Criminal.StageRaid(source, params, cb)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then
    if cb then cb(403, nil) end
    return
  end
  params = params or {}
  -- Resolve participant sources to principal IDs
  local participantPrincipalIds = {}
  if type(params.participants) == 'table' then
    for _, participantSource in ipairs(params.participants) do
      local pid = ATC.Accounts.GetPrincipalId(tonumber(participantSource))
      if pid then
        table.insert(participantPrincipalIds, pid)
      end
    end
  end
  -- Ensure lead is included
  if #participantPrincipalIds == 0 then
    table.insert(participantPrincipalIds, principalId)
  end
  apiPost('/api/v1/criminal/raids', {
    propertyId            = params.propertyId,
    initiatingAgencyId    = params.initiatingAgencyId,
    leadPrincipalId       = principalId,
    participants          = participantPrincipalIds,
    notes                 = params.notes,
  }, cb)
end

--- Start a staged raid.
--- @param raidId          string
--- @param cb              function callback(status, raid)
function ATC.Criminal.StartRaid(raidId, cb)
  apiPost('/api/v1/criminal/raids/' .. raidId .. '/start', {}, cb)
end

--- Complete a raid.
--- @param raidId          string
--- @param outcome         string   'success'|'failure'|'partial'|'aborted'
--- @param notes           string   optional
--- @param cb              function callback(status, raid)
function ATC.Criminal.CompleteRaid(raidId, outcome, notes, cb)
  apiPost('/api/v1/criminal/raids/' .. raidId .. '/complete', {
    outcome = outcome or 'success',
    notes   = notes,
  }, cb)
end

--- Abort a raid.
--- @param raidId          string
--- @param notes           string   optional
--- @param cb              function callback(status, raid)
function ATC.Criminal.AbortRaid(raidId, notes, cb)
  apiPost('/api/v1/criminal/raids/' .. raidId .. '/abort', {
    notes = notes,
  }, cb)
end

-- ── Server Events ─────────────────────────────────────────────────────────────

--- Client requests a gang operation start.
AddEventHandler('atc:criminal:operation:start:request', function(params)
  local source = source
  if type(params) ~= 'table' then return end
  ATC.Criminal.StartOperation(source, params, function(status, data)
    TriggerClientEvent('atc:criminal:operation:start:response', source, status, data)
  end)
end)

--- Client requests contraband registration.
AddEventHandler('atc:criminal:contraband:register:request', function(params)
  local source = source
  if type(params) ~= 'table' then return end
  ATC.Criminal.RegisterContraband(source, params, function(status, data)
    TriggerClientEvent('atc:criminal:contraband:register:response', source, status, data)
  end)
end)

--- Law enforcement requests contraband seizure.
AddEventHandler('atc:criminal:contraband:seize:request', function(contrabandId)
  local source = source
  if type(contrabandId) ~= 'string' then return end
  ATC.Criminal.SeizeContraband(source, contrabandId, function(status, data)
    TriggerClientEvent('atc:criminal:contraband:seize:response', source, status, data)
  end)
end)
